lib/common/s3.sh

```diff
@@ -89,3 +89,53 @@ s3_ls() {
     $S3CMD ls $S3_BUCKET/$path/ | tr -s " " | cut -d' ' -f4 | sed -e "s#^$S3_BUCKET/$path/##"
 }
 export -f s3_ls
+
+# list recursively a folder on s3 and returns list of fullpath of file objects
+# $1 - path of folder on s3
+s3_ls_recur() {
+    local folder=${1%/}; shift # remove trailing slash
+    local object
+
+    for object in `s3lsv -a -b imos-data -k $folder/`; do
+        # check if object is a folder
+        if `echo "$object" | grep -E -q '\/$'`; then
+            s3_ls_recur "$object"
+        else
+            echo "$object" | grep -o -E "$folder.*$"
+        fi
+    done
+}
+export -f s3_ls_recur
+
+# download recursively s3 objects from a given folder object
+# ex: s3_dl_recur_latest_version UWA /tmp/download_test
+# $1 - path of folder on s3
+# $2 - path of output folder where files will be downloaded keeping same structure
+s3_dl_recur_latest_version() {
+    local folder=$1; shift
+    local output_folder=$1; shift
+    [ -z "$output_folder" ] && { echo "output_folder is empty"; return 1; }
+    mkdir -p $output_folder  || return 1
+
+    # only keeping uniq filename objects
+    local file_list=`s3_ls_recur $folder | uniq | sort`
+    local object
+    local obj_info
+
+    for object in $file_list; do
+        obj_info=`s3lsv -a -b imos-data -k $object`
+        # if more than one version of same object, we reverse sort the second column(date) and take the first line
+        # i.e. the lastest object version
+        # date format is yyyy-mm-ddTHH:MM:SS
+        obj_info=`echo $obj_info | sort -b -r -k 2.1,2.4 -k 2.6,2.7 -k 2.9,2.10 -k 2.12,2.13 -k 2.15,2.16 -k 2.18,2.19 | head -1`
```

Comment:
No need to sort. `s3lsv` output is already sorted. Current version is the first line.